Support for standard in/out

jrieken opened the pull request:
This PR adds support for standard in/out (in addition to http) to Omnisharp. In contrast to #69 it's implemented as owin server which is driven by standard in. That means you run Omnisharp either with Kestrel (as a web server) or Stdio. The following changes are made with this PR
* add a new flag ```--stdio``` to Omnisharp which means read from/write to the console
* adds Omnisharp.Stdio which implements a owin server that reads single lines from Console.In, interprets them as v8-debugger style JSON message, and generates corresponding messages
* adds a special logger which writes event-packets to Console.Out when logging
* adds Omnisharp.Stdio.Tests which tests the SharedTextWriter and the StdioResponseStream

nosami commented:
Looks like a huge improvement to me. No idea how to test this thing though :)

jrieken commented:
@nosami - there are a few tests for the tricky bits (stream, shared writer) but the none for the rest. it's more or less glue code imo.

nosami commented:
Yup. Happy to merge this once you've fixed @davidfowl nitpicks

davidfowl commented:
@jrieken needs a rebase

davidfowl commented on src/OmniSharp.Stdio/StdioResponseStream.cs:
```diff
@@ -0,0 +1,187 @@
+
+            try
+            {
+                _writer.Write(",");
```
Should this really be in the Stream implementation?

jrieken replied:
do you mean the ```,``` or all of it (the rest of the object)? With the comma there is the case that when nothing has been written it's not needed

davidfowl commented on src/OmniSharp.Stdio/Services/ISharedTextWriter.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace OmniSharp.Stdio.Services
+{
+    public interface ISharedTextWriter
```
This interface feels a bit over engineered to me. Why does it take a callback?

jrieken replied:
That way the control over the lock is encapsulated, esp for error cases.

jrieken commented:
@davidfowl rebased now

jrieken commented:
@davidfowl - as requested I removed the ingenious response stream at the cost of a MemoryStream :disappointed: Also removed the stdio test project which now become pretty much useless.

davidfowl commented on src/OmniSharp.Stdio/Services/SharedTextWriter.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public Task WriteLineAsync(object value)
+        {
+            return Task.Factory.StartNew(() => WriteLine(value));
```
What's the point of this?

jrieken replied:
This is used by the logger so that it doesn't block the rest of the app

davidfowl commented on src/OmniSharp/Program.cs:
```diff
@@ -55,14 +57,19 @@ public void Main(string[] args)
              .AddCommandLine(new[] { "--server.urls", "http://localhost:" + serverPort });
 
             var serviceCollection = HostingServices.Create(_serviceProvider, config);
+            serviceCollection.AddSingleton<ISharedTextWriter, SharedConsoleWriter>();
```
This actually doesn't work.

jrieken replied:
It does - with the right workaround in ```Startup.cs```

davidfowl commented:
:shipit:

nosami commented:
:+1:
